src/engine/SCons/Node/FS.py

```diff
@@ -3243,9 +3262,127 @@ def changed_content(self, target, prev_ni):
+        if (len(binfo.bsourcesigs) + len(binfo.bdependsigs) + \
+                len(binfo.bimplicitsigs)) == 0:
+            return {}
+
+        pairs = [
+            (binfo.bsources, binfo.bsourcesigs),
+            (binfo.bdepends, binfo.bdependsigs),
+            (binfo.bimplicit, binfo.bimplicitsigs)
+           ]
+
+        m = {}
+        for children, signatures in pairs:
+            for child, signature in zip(children, signatures):
+                schild = str(child)
+                m[schild] = signature
+
+
+
+        # store this info so we can avoid regenerating it.
+        binfo.dependency_map = m
+        return m
+
+    def _get_previous_signatures(self, dmap):
+        """
+        Return a list of corresponding csigs from previous
+        build in order of the node/files in children.
+
+        Args:
+            self - self
+            dmap - Dictionary of file -> csig
+            children - List of children
+
+        Returns:
+            List of csigs for provided list of children
+        """
+        prev = []
+
+
+
+        # First try the simple name for node
+        c_str = str(self)
+        if os.altsep:
+            c_str = c_str.replace(os.sep, os.altsep)
+        df = dmap.get(c_str)
+        if not df:
+            print("No Luck1:%s"%c_str)
+            try:
+                # this should yield a path which matches what's in the sconsign
+                c_str = self.get_path()
+                if os.altsep:
+                    c_str = c_str.replace(os.sep, os.altsep)
+
+                df = dmap.get(c_str)
+                if not df:
+                    print("No Luck:%s"%[str(s) for s in self.find_repo_file()])
+                    print("       :%s"%self.rfile())
+
+            except AttributeError as e:
+                import pdb;
+                pdb.set_trace()
+
+
+        if df:
+            return df
+        else:
+            print("CHANGE_DEBUG: file:%s PREV_BUILD_FILES:%s" % (c_str, ",".join(dmap.keys())))
+            pass
+
+        return None
+
+    def changed_timestamp_then_content(self, target, prev_ni, node=None):
+        """
+        Used when decider for file is Timestamp-MD5
+
+        NOTE: If the timestamp hasn't changed this will skip md5'ing the
+              file and just copy the prev_ni provided.  If the prev_ni
+              is wrong. It will propagate it.
+              See: https://github.com/SCons/scons/issues/2980
+        
+        Args:
+            self - dependency
+            target - target
+            prev_ni - The NodeInfo object loaded from previous builds .sconsign
+
+        Returns: 
+            Boolean - Indicates if node(File) has changed.
+        """
+        if node is None:
+            raise DeciderNeedsNode(self.changed_timestamp_then_content)
+
+        # Now get sconsign name -> csig map and then get proper prev_ni if possible
+        bi = node.get_stored_info().binfo
+        try:
+            dependency_map = bi.dependency_map
```

Comment:
I would probably use "hasattr" at this place instead. For me, it tells a bit more explicit what's actually going on...

Reply:
I"m going to leave as is, as theres much other code written similarly.
In general try/except is faster than hasattr, depending if you hit the common case in the try.